src/charm.py

```diff
@@ -369,7 +370,7 @@ def _external_hostname(self) -> str:
         return self.config["web_external_url"] or f"{self.app.name}"
 
     @property
-    def _remote_write_address(self) -> str:
+    def _remote_write_address(self) -> Optional[str]:
```

Comment:
This is bad practice for the following reason -  An empty string is false in a boolean context. Hence the return type does not need to be an optional string, it can always be a string. This ensures every method always has a single, unique, well defined return type. Failure to ensure single unique return types in general is bad practice because it prevents composition of functions i.e f(g(x)) etc.

Reply:
I agree on this one.

Reply:
As do I. But this means it should return `""`, not `None`, and the constructor for `PrometheusRemoteWriteProvider` should take `endpoint_address: Optional[str] = ""`. 

It doesn't otherwise change the logic.

Reply:
To emphasize Ryan's comment: this change is correct as far as typing goes because of what is returned:
```
return self._external_hostname if self.model.get_relation("ingress") else None
```
If you'd like, I can change it to `else ""`.
However, in this particular case, returning None is correct semantically and "" is wrong semantically: if there is no ip it is most likely you would not want to compose.

Reply:
Python is dynamically typed. Embrace it!

Reply:
This is a false understanding of "Dynamic Typing". Dynamic Typing is effectively employed when you use alternatives types that have the same/similar interface (behaviour) a.k.a duck typing. However if you pass a `NoneType` instead of a `string`, these two types do not have the same interface/behaviour. If your return type was an `Iterable` and you returned a either a list or a string that would still be okay though and an ideal design.

Reply:
To elaborate on my points above consider what happens if some other methods that invokes this one writes something like this
```
if len(returned_address) > 0:
    blah blah blah 
```
If you always returned a string this would not blow up. It does blow up because you return `NoneType`.
On the other hand if the invoking method did
```
if returned_address:
    blah blah blah
```
It is indeed fine even with NoneType but it will also be fine with empty string instead of NoneType.
Which solution "makes your code easy to use".

Reply:
Captured in https://github.com/canonical/prometheus-k8s-operator/issues/206 as this change would be outside the scope of this PR.